src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+{
+
+  /// <summary>
+  /// The type of external data source
+  /// </summary>
+  public enum DataSourceType: int {
+    /// <summary>
+    /// No datas ource is specified. This is the default state.
+    /// </summary>
+    Null = 0,
+    
+    /// <summary>
+    /// Data source is a comma separated value (CSV) file.
+    /// </summary>
+    Csv = 1,
+    
+    /// <summary>
+    /// Data source is an XML document
+    /// </summary>
+    Xml = 2,
+    
+    /// <summary>
+    /// Data source is a database
+    /// </summary>
+    Db = 4
+  };
+  
+  /// <summary>
+  /// Common amounts of data rows to collect
+  /// </summary>
+  public enum DataRowCount: int {
+    /// <summary>
+    /// Read all rows
+    /// </summary>
+    All = 0
+  }
+  
+  
+  /// <summary>
+  /// Specifies an external data source for test methods. This is evaulated at compile time.
+  /// </summary>
+  [AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
+  public class DataSourceAttribute : NUnitAttribute, ITestBuilder, IImplyFixture
+  {
+    private NUnitTestCaseBuilder _builder = new NUnitTestCaseBuilder();
+
+    #region Constructors
+
+    /// <summary>
+    /// Constructs the object responsible for obtaining test data.
+    /// </summary>
+    /// <param name="dataType">The type of data provider</param>
+    /// <param name="sourceName">The connection string for the data source. It could be a filename or DB connection string.</param>
+    /// <param name="rowCount">The number of rows to collect from the data source. By default it is all rows</param>
+    public DataSourceAttribute(DataSourceType dataType, string sourceName, int rowCount = (int)DataRowCount.All)
+    {
+      this.SourceName = sourceName;
+      this.DataType = dataType;
+      this.DataRows = rowCount;
+    }
+    
+    #endregion
+
+    #region Properties
+
+    /// <summary>a
+    /// The name of a file or connection string which contains the data
+    /// </summary>
+    public string SourceName { get; private set; }
+
+    /// <summary>
+    /// The data input type
+    /// </summary>
+    public DataSourceType DataType {get; private set;}
+
+    /// <summary>
+    /// The number of rows to read from the source. Zero takes a special meaning to be all rows. This is the default.
+    /// </summary>
+    public int DataRows { get; set;}
+    
+    #endregion
+    
+    #region ITestCaseSource Members
+    
+    /// <summary>
+    /// Returns a set of test case parameters for use as arguments
+    /// to a parameterized test method.
+    /// </summary>
+    /// <param name="method">The method for which data is needed.</param>
+    /// <returns></returns>
+    private IEnumerable<TestCaseParameters> GetTestCasesFor(IMethodInfo method)
+    {
+      List<TestCaseParameters> data = new List<TestCaseParameters>();
+
+      try
+      {
+        IEnumerable source = GetDataFromSource(method);
+
+        if(source != null)
+        {
+          #if NETCF
```

Comment:
I've only tested on the desktop. I took an existing attribute to learn the structure and then modified it accordingly. I'm trying to use as many generic calls as I can for it to be compatible on different framework versions.